### ISSUE TITLE
Kodi support questions

### ISSUE THREAD

**Issue by svenfoo** [benzea/gnome-screencast#81]
_Please excuse me filing an issue to get in contact and asking questions. Feel free to direct me to a more appropriate channel._

I'd like to be able to cast my GNOME desktop to the Raspberry Pi running Kodi. It seems that gnome-network-displays might provide the GNOME side of things, but I haven't been able to use it with Airplay enabled on Kodi. I'd like to figure out what needs to be done on the Kodi side to make this work. So here are some questions:

1. Is it possible to use gnome-network-displays without WiFi P2P? I already have the Raspberry Pi connected to my local network via ethernet, and I my notebook running GNOME has a good WiFi connection. I don't think that a direct wireless connection would yield more bandwidth and I would rather avoid the hassle of teaching Kodi to do P2P

2.  What protocols does gnome-network-displays support and what protocol would you suggest to be used on the Kodi side? As far as I know Kodi already has Airplay support, but I would actually prefer to use a non-proprietary protocol.

My goal is not to cast some video or game running on the notebook, it is rather about sharing the screen in order to allow multiple people to participate in coding or text editing. So frame rate is not much of an issue, good readability (few artefacts) would be more important.

Thanks,
Sven

**Comment by benzea** [comment]
1. It is supportable in principle, but not implemented. Note that you really want a direct WiFi connection rather than forwarding the traffic via the router (which doubles the required air-time). If both devices are connected to the same AP, then this is usually solved using TDLS. I don't know whether we are doing TDLS correctly though (I kind of hope it just happens magically, but have not done the research).
2. Currently only WiFi Display revision 1 is supported.

I can't really help much right now, I don't know what protocols Kodi supports. That said, it might be easier to support further sink types in gnome-network-displays. The internal API has been abstracted specifically to allow this.

**Comment by alex-eri** [comment]
For Kodi is better to stream `RTSP` video with `VLC` or `OBS`.

like this
https://superuser.com/a/947141/239779

there is some latency, about 200ms

OBS Studio is better for you i think

**Comment by alex-eri** [comment]
@benzea If DLNA/UPNP protocols will be supported - it will be good. Kodi, TVs and many STB will be sopported. But latency of rtsp is not so good as miracast.

**Comment by benzea** [comment]
Well, WiFi-Display is basically RTSP. So that is likely easier to do on top of what exists right now.

**Comment by svenfoo** [comment]
Thanks for the answers. Still not convinced that I want to have a direct WiFi connection as with my setup there is only one client on the WiFi. The Raspberry Pi running Kodi is connected over ethernet, so available air-time should be comparable.

VLC seems to be a good choice for sending a video stream to Kodi, but I am specifically interested in mirroring my desktop screen and gnome-network-displays seems to have nice integration with the GNOME desktop for doing that.

With your hint on WiFi-Display I have been able to find some more information, thanks a lot. It seems that https://github.com/albfan/miraclecast might be able to provide everything that's needed on the KODI side.

**Comment by alex-eri** [comment]
Oh, I forget... Kodi can already connect with RTSP on debug display (` NETWORK_DISPLAYS_DUMMY=1`)

Start gnome-network-displays with

`
$  NETWORK_DISPLAYS_DUMMY=1 gnome-network-displays
`

Create m3u playlist and put it to Kodi

```
#EXTINF:0,WFD Stream
rtsp://192.168.1.33:7236/wfd1.0

```

**Comment by benzea** [comment]
Yeah, so adding code to discover a kodi and then put gnome-network-displays into a similar mode would likely be completely reasonable.

Hrm, eventually, one probably wants to make sure the wrong people don't connect … like right now, I guess someone in the same network might be able to grab the video stream just by connecting.

**Comment by svenfoo** [comment]
Oh, wow, this kind of works. However the connection is dropped after a few seconds. It seems that the audio is to blame:
```
** (gnome-network-displays:582359): WARNING **: 09:34:41.451: WfdClient: Got OPTIONS before getting reply querying WFD support; assuming normal RTSP connection.
Warning from pulseaudio-source: Can't record audio fast enough
```
The log on the kodi side:
```
2020-01-20 09:34:44.637 T:3011576416  NOTICE: VideoPlayer::OpenFile: /storage/videos/network-display.strm
2020-01-20 09:34:44.639 T:2561651568  NOTICE: Creating InputStream
2020-01-20 09:34:44.639 T:2561651568  NOTICE: Creating Demuxer
2020-01-20 09:34:54.752 T:2561651568  NOTICE: Opening stream: 0 source: 256
2020-01-20 09:34:54.752 T:2561651568  NOTICE: Creating video codec with codec id: 27
2020-01-20 09:34:54.775 T:2561651568  NOTICE: Creating video thread
2020-01-20 09:34:54.776 T:2872443760  NOTICE: running thread: video_thread
2020-01-20 09:34:54.776 T:2561651568  NOTICE: Opening stream: 1 source: 256
2020-01-20 09:34:54.776 T:2561651568  NOTICE: Finding audio codec for: 86018
2020-01-20 09:34:54.782 T:2561651568  NOTICE: CDVDAudioCodecFFmpeg::Open() Successful opened audio decoder aac
2020-01-20 09:34:54.782 T:2561651568  NOTICE: Creating audio thread
2020-01-20 09:34:54.783 T:2498462576  NOTICE: running thread: CVideoPlayerAudio::Process()
2020-01-20 09:34:54.793 T:2498462576  NOTICE: Creating audio stream (codec id: 86018, channels: 2, sample rate: 48000, no pass-through)
2020-01-20 09:34:55.369 T:2872443760 WARNING: CRenderManager::WaitForBuffer - timeout waiting for buffer
2020-01-20 09:34:55.944 T:2989884272 WARNING: ActiveAE - large audio sync error: 1449.336973
2020-01-20 09:34:56.112 T:2989884272 WARNING: ActiveAE - large audio sync error: 1859.487659
2020-01-20 09:34:56.532 T:2989884272 WARNING: ActiveAE - large audio sync error: 1561.043314
2020-01-20 09:35:17.546 T:2498462576  NOTICE: CVideoPlayerAudio::Process - stream stalled
2020-01-20 09:35:17.565 T:2561651568  NOTICE: CVideoPlayer::OnExit()
2020-01-20 09:35:17.565 T:2561651568  NOTICE: VideoPlayer: eof, waiting for queues to empty
2020-01-20 09:35:17.565 T:2561651568  NOTICE: Closing stream player 1
2020-01-20 09:35:17.566 T:2561651568  NOTICE: CDVDMessageQueue(audio)::WaitUntilEmpty
2020-01-20 09:35:17.566 T:2561651568  NOTICE: Waiting for audio thread to exit
2020-01-20 09:35:17.566 T:2498462576   ERROR: Got MSGQ_ABORT or MSGO_IS_ERROR return true
2020-01-20 09:35:17.566 T:2498462576  NOTICE: thread end: CVideoPlayerAudio::OnExit()
2020-01-20 09:35:17.566 T:2561651568  NOTICE: Closing audio device
2020-01-20 09:35:17.567 T:2561651568  NOTICE: Deleting audio codec
2020-01-20 09:35:17.568 T:2561651568  NOTICE: Closing stream player 2
2020-01-20 09:35:17.568 T:2561651568  NOTICE: CDVDMessageQueue(video)::WaitUntilEmpty
2020-01-20 09:35:17.685 T:2561651568  NOTICE: waiting for video thread to exit
2020-01-20 09:35:17.686 T:2872443760   ERROR: Got MSGQ_ABORT or MSGO_IS_ERROR return true
2020-01-20 09:35:17.686 T:2872443760  NOTICE: thread end: video_thread
2020-01-20 09:35:17.686 T:2561651568  NOTICE: deleting video codec
2020-01-20 09:35:17.774 T:3011576416  NOTICE: CVideoPlayer::CloseFile()
2020-01-20 09:35:17.826 T:3011576416  NOTICE: VideoPlayer: waiting for threads to exit
2020-01-20 09:35:17.826 T:3011576416  NOTICE: VideoPlayer: finished waiting
2020-01-20 09:35:17.826 T:3011576416  NOTICE: CVideoPlayer::CloseFile()
2020-01-20 09:35:17.826 T:3011576416  NOTICE: VideoPlayer: waiting for threads to exit
2020-01-20 09:35:17.827 T:3011576416  NOTICE: VideoPlayer: finished waiting
```

Is there an easy way to create a video-only stream? For the desktop sharing use case we do not need audio at all.

**Comment by benzea** [comment]
Uh, does it disconnect after about 30s?

It could be that the keep-alive is not working for some reason. That seems more likely than the audio related warning (which generally means that the video encoder took too long for a frame).

The keep-alive is an empty `GET_PARAMS` call that is sent to the sink (i.e. Kodi) every 25s. If it does not reply (or do something else to keep the connection alive), then the RTSP connection times out and the stream is killed.

**Comment by svenfoo** [comment]
You are most probably right. It disconnects almost immediately after the debug message **WfdClient: Doing keep-alive** appeared.

**Comment by svenfoo** [comment]
It looks like Kodi uses **ffmpeg** to handle the RTMP stream. I can use ffmpeg directly instead of using Kodi:

```
ffmpeg -loglevel debug -i rtsp://localhost:7236/wfd1.0 -acodec libmp3lame -ar 22050 -vcodec copy -y -f flv
```

Using this command-line the connection with gnome-network-displays is first created, but it disconnects very soon after the debug message **WfdClient: Doing keep-alive** appears. So the problem doesn't seem to be specific to Kodi. It looks more like gnome-network-displays and ffmpeg do not like each other.

ffmpeg version 4.2.2

**Comment by svenfoo** [comment]
I spent a little more time trying to debug this and I have a couple of questions now. @benzea, maybe you could find a few minutes and attempt to answer them?

From my understanding of the RTSP session timeout, I would describe it like this:

There is a timeout of 30 seconds on the session. If the client does not interact with the server for longer than this timeout, then the server decides that the client has become inactive, stops sending the streams and ends the session. So the client needs to regularly send commands. Typically it will send a GET_PARAMETER call with an empty parameter list, this is sort of a no-op command. To summarize this, unless the client sends keep-alive commands to the server, the server will stop streaming.

In gnome-network-displays there is code in the **WfdClient** that sends such an empty  GET_PARAMETER call every 25 seconds. But what's the purpose of this code? WfdClient is the server-side representation of the streaming client. So that code sends a keep-alive from the server to the client, doesn't it? That seems wrong to me, and I wonder what the purpose of this code is.

So I went ahead and commented out https://github.com/benzea/gnome-network-displays/blob/8a7bb6b993e9e8ffbbcbcc34c569e953af682d57/src/wfd/wfd-client.c#L373
That basically removes the whole keep-alive functionality in WfdClient. And voilà, without this the connection from Kodi, or ffmpeg, becomes stable.

I've had a look at the ffmpeg RTSP client code, and there's a keep-alive implementation there that makes sure that there is interaction with the server before the session times out. It uses exactly the described mechanism of sending an empty GET_PARAMETER call unless there has been other interaction.

So in my opinion the keep-alive code in WfdClient is unnecessary and seems to cause problems. If you agree then I would open a pull request that removes the timeout and callback. But then I guess the code has been added for a reason...

**Comment by benzea** [comment]
It is mandatory in the WiFi-Display specification.

Did you test this with the latest gnome-network-displays? There was a bugfix where we don't do a keep-alive if the response does not contain any session.

Also, can you check how ffmpeg responds to our keep-alive?

**Comment by svenfoo** [comment]
I've tested with gnome-network-displays as shipped with Fedora 31 and with a binary built today from the master branch (8a7bb6b993). Both show the problem. After commenting out the aforementioned line the problem goes away.

I haven't yet checked in detail how ffmpeg deals with the keep-alive call from the server.

**Comment by benzea** [comment]
It would be enough to just get a tcpdump (e.g. using wireshark) for the RTSP connection. It should be obvious what is happening from that.

**Comment by svenfoo** [comment]
[rtsp-ffmpeg.txt](https://github.com/benzea/gnome-network-displays/files/4119406/rtsp-ffmpeg.txt)

**Comment by benzea** [comment]
haha, nice one, it just errors out.

Well, that makes it simple, we could just immediately do one of these at the start, and if it does not work, we stop doing it.

Or, actually, we run into a protocol difference earlier anyway (you get a warning becaues of that). Really, at that point we could disable our own keep-alive action …

**Comment by svenfoo** [comment]
I assume you are referring to this warning:
`
WfdClient: Got OPTIONS before getting reply querying WFD support; assuming normal RTSP connection.
`

So we could flag the connection as "normal RTSP connection" then and disable the keep-alive, right? Will look into providing a patch that implements this.

**Comment by benzea** [comment]
Forgot to close the issue (would have automatically if you had written `Fixes: #81` or similar).
